src/stream/src/common/table/state_table.rs

```diff
@@ -46,16 +46,23 @@ use risingwave_storage::row_serde::value_serde::ValueRowSerde;
 use risingwave_storage::store::{
     LocalStateStore, NewLocalOptions, PrefetchOptions, ReadOptions, StateStoreIterItemStream,
 };
+use risingwave_storage::table::merge_sort::merge_sort;
 use risingwave_storage::table::{compute_chunk_vnode, compute_vnode, get_second, Distribution};
 use risingwave_storage::StateStore;
 use tracing::{trace, Instrument};
 
 use super::watermark::{WatermarkBufferByEpoch, WatermarkBufferStrategy};
 use crate::cache::cache_may_stale;
+use crate::common::cache::{StateCache, StateCacheFiller};
+use crate::common::table::state_table_cache::StateTableWatermarkCache;
 use crate::executor::{StreamExecutorError, StreamExecutorResult};
 
 /// This num is arbitrary and we may want to improve this choice in the future.
 const STATE_CLEANING_PERIOD_EPOCH: usize = 300;
+/// This num is arbitrary and we may want to improve this choice in the future.
+const WATERMARK_CACHE_ENTRIES: usize = 1024;
```

Comment:
Personally, I tend to use a much smaller size (such as 16) because deletes/updates should be very rare in watermark-enabled operators. Mostly are purely append-only input, in which case `cache_size = 1` would be enough.

Reply:
Please correct me if I made wrong assumption

Reply:
Makes sense. Let's do 16 to start with.